feat: add optimizeDeps.esbuildOptions

nihalgonsalves opened the pull request:
<!-- Thank you for contributing! -->

### Description

This adds the ability to customise the esbuild options during the optimiser phase, including plugins.

There are some libraries that esbuild throws an error on, and I'd really like to have this functionality – we're stuck on a Vite conversion because of one or two packages – with plain esbuild, we're able to use plugins to work around small errors.

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

### Additional context

See #3124

This is inspired by the conversation here: https://github.com/vitejs/vite/pull/2886#discussion_r608410494

The discussion is about where the build options should live. I feel like since `optimizeDeps` is an entirely different build phase than the esbuild transform options, the settings should go into `config.optimizeDeps`.

### Questions

- Other than `define` and `entryPoints`, should any other options be disallowed? For example, perhaps overriding `bundle`, `format`, or `outdir` shouldn't be allowed.
- Should the old fields be deprecated, or just left as additional options?

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

---

Closes #2886

nihalgonsalves committed:
feat: add optimizeDeps.esbuildOptions

qnp commented:
Thanks for your PR, related to mine. In my work I have also to include eslint plugins in this scan phase : https://github.com/vitejs/vite/blob/7bcbe907ce13e1fc888a92a331db774f277fde5e/packages/vite/src/node/optimizer/scan.ts#L96 

1. I'm not 100% sure on whether all passed esbuild options through `config.optimizeDeps` should be also present here.

2. Regarding your 1st question, I think the Vite core team should clarify us their philosophy. To me, I would keep Vite as flexible as possible, with warnings on what should and what should not be done in the documentation.

3. Regarding your second question, your strategy to deprecate these redundant fields seems right.

qnp commented:
Yet, I'm not convinced about the choice to place `esbuildOptions` into `optmimizeDeps`. For me, it is not clear to have `config.optmimizeDeps` in one side with some esbuild _build_ options in it, and `config.esbuild` in another side with some esbuild _transform_ options. We should work on clarifying this API.

Keeping your idea in mind, we should also deprecate `config.esbuild` options in favor of:
```
config = {
  transformDeps: {
    include: ...,
    exclude: ..., // deprecate ??
    jsxInject: `import React from 'react'`,
    esbuildOptions: {
      jsxFactory: 'h',
      jsxFragment: 'Fragment',
      /* any other esbuild _transform_ option */
    }
  }
}

patak-js commented:
@nihalgonsalves we discussed with Evan, this PR is a good first step towards allowing users to patch the optimizer when needed. We can continue talking about other further improvements in https://github.com/vitejs/vite/issues/3124, but for the time being, I think that this is enough given the tradeoffs regarding increasing Vite's API surface. Thanks again for pushing for these changes.

@qnp about https://github.com/vitejs/vite/pull/2991#issuecomment-820229102, I agree with you that this is a better API for `config.esbuild`. Let's discuss about this in a separate PR though. It shouldn't be `transformDeps` as this is applied to your own sources too, maybe just `transform`. My initial thoughts on this are that we should wait a bit more to change this though.

Regarding https://github.com/vitejs/vite/pull/2991#issuecomment-820219153,

> I'm not 100% sure on whether all passed esbuild options through `config.optimizeDeps` should be also present [here](https://github.com/vitejs/vite/blob/7bcbe907ce13e1fc888a92a331db774f277fde5e/packages/vite/src/node/optimizer/scan.ts#L96).

I see two options here. We could add a new `optimizeDeps.scan.esbuildOptions`, to be discussed in another PR. Or we could review if it makes sense to use some of `optimizeDeps.esbuildOptions` in https://github.com/vitejs/vite/blob/1e785d1af54ae5e305dd1bef9af513f2c3a91ad3/packages/vite/src/node/optimizer/scan.ts#L94-L105
I am not sure at this point what is the best strategy. @nihalgonsalves @antfu could you help to review this?

> Regarding your 1st question, I think the Vite core team should clarify us their philosophy. To me, I would keep Vite as flexible as possible, with warnings on what should and what should not be done in the documentation.

At this point, Vite shouldn't tie itself too hard to the internal libraries it is using. For example, IMO it would have been better to name `build.watch` like `build.chokidarOptions` because it is easier to deprecate and move to a different lib later. Better than that is having our own API that doesn't depend on the used tools. For example `optimizeDeps.exclude`, that can be rewired if we ever need to change it.

> Regarding your second question, your strategy to deprecate these redundant fields seems right.

I agree here about `keepNames` since that is more like an esbuild only detail. But `optimizeDeps.exclude` is a Vite feature. I think we should merge it with `optimizeDeps.esbuildOptions.external` instead of deprecating it.

nihalgonsalves committed:
fix: remove unsupported @see to silence warning

(API extractor warning)

nihalgonsalves committed:
test: add esbuild-plugin integration test

nihalgonsalves commented on packages/vite/src/node/config.ts:
```diff
@@ -381,10 +385,18 @@ export async function resolveConfig(
+    createResolver,
+    optimizeDeps: {
+      ...config.optimizeDeps,
+      esbuildOptions: {
+        keepNames: config.optimizeDeps?.keepNames,
+        external: config.optimizeDeps?.exclude,
+        ...config.optimizeDeps?.esbuildOptions
+      }
```
I went for defaulting to the top-level options, but letting users replace them. Merging exclude with external could be surprising for users, if you set `external`, it should be respected.

patak-js replied:
I think it may be better to include `external` in the list of excluded options. The rationale is that we may use `optimizeDeps.exclude` for something else (maybe logging, or some future feature). Since you can already achieve with `exclude` what you get with `external`, looks to me that this is the safer choice.

nihalgonsalves replied:
Yeah, makes sense. Removed it.

nihalgonsalves commented on packages/vite/src/node/optimizer/index.ts:
```diff
@@ -47,9 +47,28 @@ export interface DepOptimizationOptions {
+   */
+  esbuildOptions?: Omit<
+    EsbuildBuildOptions,
+    | 'bundle'
+    | 'write'
+    | 'watch'
+    | 'outdir'
+    | 'outfile'
+    | 'outbase'
+    | 'outExtension'
+    | 'metafile'
```
Added this because I don't believe changing these would have a valid use case. Vite depends on bundling, the metafile, the output settings, and the build mode (i.e. watch = false).

Shinigami92 replied:
Should we add a comment within the code about why they are omitted?

nihalgonsalves replied:
I like the inline documentation of what's omitted, but I added a comment.

e.g.:

![image](https://user-images.githubusercontent.com/10807310/116794696-4eaf5e00-aacf-11eb-8ecc-e9679193cdf9.png)

nihalgonsalves commented on packages/vite/src/node/optimizer/index.ts:
```diff
@@ -233,20 +252,25 @@ export async function optimizeDeps(
 
   const start = Date.now()
 
+  const { entryPoints = [], plugins = [], ...esbuildOptions } =
+    config.optimizeDeps?.esbuildOptions ?? {}
+
   const result = await build({
-    entryPoints: Object.keys(flatIdDeps),
+    entryPoints: [...entryPoints, ...Object.keys(flatIdDeps)],
```
Does it make sense to merge the user's `entryPoints`? Or should we simply add it to the excluded settings?

patak-js replied:
What would be the use case of adding extra `entryPoints` here? There is logic needed to flat ids that wouldn't be applied if the user has control of them. Maybe we should also omit this option (it could be added later also in case we find a use case)

nihalgonsalves replied:
Yeah, makes sense. I removed it.

nihalgonsalves commented on packages/vite/src/node/optimizer/scan.ts:
```diff
@@ -85,33 +84,23 @@ export async function scanImports(
+
+  await build({
+    write: false,
+    entryPoints: [...entryPoints, ...entries],
```
Same question here about merging `entryPoints`.

Additionally, I moved it from multiple builds to a single one – should be marginally faster. Have I not considered a reason to do it separately?

nihalgonsalves reviewed (commented):
@patak-js All done.

About the question about dep-scanning: this PR now also uses the same custom options for dep scanning. I think that if you're trying to patch something or change behaviour, you'd want your settings to apply to both the pre-scan and the bundling.

Some explanations of the choices I made below:

nihalgonsalves committed:
fix: rollback scan multiple entrypoints at once

nihalgonsalves committed:
test: adapt esbuild dep test to also work in build

nihalgonsalves committed:
test: fix selector & remove log

nihalgonsalves commented:
About the question about dep-scanning: this PR now also uses the same custom options for dep scanning. I think that you're trying to patch something or change behaviour, you'd want your settings to apply to both the pre-scan and the bundling.

nihalgonsalves committed:
test: attempt a cross platform path match?

Just a guess since I’m not developing on Windows, but perhaps it doesn’t match on the CI due to the back slash on Windows

nihalgonsalves committed:
refactor(optimize-deps): omit esbuild entrypoints

nihalgonsalves committed:
refactor(opt-deps): omit esbuild’s external option

in favour of setting exclude directly.

patak-js commented on packages/vite/src/node/optimizer/scan.ts:
```diff
@@ -85,34 +84,28 @@ export async function scanImports(
     )
   )
 
-  try {
-    emptyDir(tempDir)
-    fs.rmdirSync(tempDir)
-  } catch (err) {
-    if (err.code !== 'ENOENT') {
-      throw err
-    }
-  }
-
```
Nice! This is great, but I think we should push it to another PR. Just in case there are issues, I think this PR should only be about adding the esbuild options, so we can better track this with git blame in case there are regressions. The refactoring to log deprecated options also looks good, I'm debating if we should also do it in another PR (maybe this one is fine)

nihalgonsalves replied:
Yeah, fair enough, removed the tmpdir stuff. I think the log stuff is fine :)

nihalgonsalves replied:
=> #3231

Shinigami92 commented on packages/vite/src/node/config.ts:
```diff
@@ -402,69 +413,88 @@ export async function resolveConfig(
-      chalk.yellow.bold(
-        `(!) "build.base" config option is deprecated. ` +
-          `"base" is now a root-level config option.`
+      chalk.bold.yellow(
```
**nitpick:** Let's switch this back to `chalk.yellow.bold` so the diff gets smaller

nihalgonsalves replied:
```suggestion
      chalk.yellow.bold(
```

Shinigami92 commented on packages/vite/src/node/config.ts:
```diff
@@ -402,69 +413,88 @@ export async function resolveConfig(
   }
 
   // TODO Deprecation warnings - remove when out of beta
-  if (config.build?.base) {
+
+  const logDeprecationWarning = (
```
**question:** Why are we updating code for deprecation?
Also what is this `beta`? From when is it?

Shinigami92 replied:
Ah okay... seeing now that you created this to solve all the duplication below
But what has it to do with the goal of the PR?
Do we need to change the git blame annotations for that in this PR?
IMO it's more worth about thinking to remove this if it's for deprecation :thinking:

nihalgonsalves replied:
Not sure about when these were intended to be removed - but since they're still part of the API, we couldn't remove them entirely in a minor version.

I think it's okay to do a bit of refactoring like this in a PR, but don't have too strong an opinion here - I can remove it if you like.

Shinigami92 replied:
I will let @patak-js decide :slightly_smiling_face:

nihalgonsalves replied:
=> https://github.com/vitejs/vite/pull/2991#discussion_r624555056

Shinigami92 replied:
Haha ^^ didn't read the discussions from you and patak before
But seems we had kinda same thoughts :sweat_smile:

patak-js replied:
I also prefer if they are separated PRs, mainly because this is could be an interesting PR to reference later. But the refactoring is contained. If @antfu reviews the PR and asks the same question, it is a clear indication that it better to separate them.

nihalgonsalves replied:
PRs are cheap, here you go: #3232

patak-js commented:
@nihalgonsalves I tested locally and found several spurious `console.log`. They are also in the CI https://github.com/vitejs/vite/pull/2991/checks?check_run_id=2483662241#step:11:11.

Shinigami92 commented on packages/vite/src/node/optimizer/index.ts:
```diff
@@ -47,9 +47,29 @@ export interface DepOptimizationOptions {
+   *
+   * - `plugins` are merged with Vite's dep plugin
+   * - `keepNames` takes precedence over the deprecated `optimizeDeps.keepNames`
+   * - `external` is omitted, use Vite's optimizeDeps.exclude option
```
```suggestion
   * - `external` is omitted, use Vite's `optimizeDeps.exclude` option
```

Shinigami92 commented on packages/vite/src/node/optimizer/index.ts:
```diff
@@ -47,9 +47,29 @@ export interface DepOptimizationOptions {
+   * - `keepNames` takes precedence over the deprecated `optimizeDeps.keepNames`
+   * - `external` is omitted, use Vite's optimizeDeps.exclude option
+   *
+   * https://esbuild.github.io/api
```
This make the link clickable in VSCode

```suggestion
   * [esbuild.github.io/api](https://esbuild.github.io/api)
```

nihalgonsalves replied:
Odd, it's clickable for me in any case

Shinigami92 replied:
Checked it and you are right. Strange, thought it doesn't work :shrug:

Shinigami92 commented on packages/vite/src/node/optimizer/index.ts:
```diff
@@ -47,9 +47,29 @@ export interface DepOptimizationOptions {
+    | 'metafile'
+  >
+  /**
+   * @deprecated use esbuildOptions.keepNames
```
```suggestion
   * @deprecated use `esbuildOptions.keepNames`
```

nihalgonsalves committed:
test: silence expected warnings in resolveConfig test

nihalgonsalves committed:
style: apply minor suggestions

Co-authored-by: Shinigami <chrissi92@hotmail.de>

nihalgonsalves commented:
> I tested locally and found several spurious console.log. They are also in the CI #2991 (checks).

@patak-js Oops, forgot about those. They're due to the test for the keepNames deprecation handling. I silenced them now

nihalgonsalves committed:
feat(scan): support custom esbuild options

nihalgonsalves commented:
Tried it out on a project, 271 dependencies to pre-bundle: 5600-5700ms before, 5300-5400ms after. (Though I did a statistically insignificant amount of runs, so take that with a grain of salt)

nihalgonsalves committed:
docs: add comment about omitted options

nihalgonsalves committed:
refactor: deprecate keepNames, not exclude

qnp commented:
Thanks for the work, LGTM
